### ISSUE TITLE
Narrow and explicit RBAC

### ISSUE THREAD

**Issue by maanur** [kubeflow/katib#1639]
/kind feature

I am trying to deploy Katib in an enterprise environment and have some hard time explaining Katib's requested RBAC
 rules. It would be much appreciated if ClusterRole explicitly declared only necessary verbs for each individual resource.

Look at this:
``` yaml
rules:
  - apiGroups:
      - ""
    resources:
      - configmaps
      - serviceaccounts
      - services
      - events
      - namespaces
      - persistentvolumes
      - persistentvolumeclaims
      - pods
      - pods/log
      - pods/status
    verbs:
      - "*"
```
Full access for Namespaces, Roles and RoleBindings effectively gives Katib unconstrained privileges to do anything it wants. This is plainly unacceptable in my case.

I suggest that ClusterRoles would be narrow and explicit like that:
```
rules:
  - apiGroups:
      - ""
    resources:
      - namespaces
    verbs:
      - get
      - list
      - watch
 - apiGroups:
      - ""
    resources:
      - roles
    verbs:
      - get
      - list
    ... e.t.c
```

E.g. let's get rid of those stars in rules.verbs and olso remove unnecessary verbs from there

**Comment by maanur** [comment]
I'll look into the issue, but I may need help with testing because of lack of free time.
Firstly, I'm going to look at kubebuilder markers and try making them work, thus getting rid of "stars" in ClusterRoles.

After that I'd be willing to switch the operator's scope to multiple namespaces instead of a whole cluster, as the current level of privileges is not necessary in fact (remember `katib-metricscollector-injection` label on tracked namespaces). That would be a breaking change, so I'll make a separate proposal for thorough consideration.

**Comment by andreyvelich** [comment]
That would be great, thank you @maanur!

> After that I'd be willing to switch the operator's scope to multiple namespaces instead of a whole cluster

Are you referring to this issue: https://github.com/kubeflow/katib/issues/1295 ?
Which runs Katib controller without cluster role access ?

**Comment by maanur** [comment]
> Are you referring to this issue: #1295 ?\nWhich runs Katib controller without cluster role access ?

Yes, partially. I suppose the particular problem of deploying multiple controllers can be solved by OLM and its OperatorGroups, but that can be considered a separate feature.

**Comment by andreyvelich** [comment]
@maanur I think that also requires additional proposal on how we want to implement this.
Currently, we don't have solution on how to enable our Webhooks on namespace mode.

Let's start with removing redundant access from the ClusterRole, after that we can discuss about role-based access.

**Comment by juliusvonkohout** [comment]
I have fixed the rules in https://github.com/kubeflow/katib/pull/1768 and documented the remaining problems.

**Comment by juliusvonkohout** [comment]
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: katib-controller
# We only need the verb create and not delete because all things are owned transitively by the
# experiment, so everything is cascade-deleted on experiment deletion.
# Early stopping currently creates a serviceaccount, role and rolebinding to manage trials and trails/status.
# The roles and rolebinding creation is the most important potential security issue now.
# Via new roles, rolebindings and pods in kubernetes system namespaces the cluster 
# could become compromised if katib-controller is compromised.
rules:
  - apiGroups:
    - ""
    - batch
    - apps
    - rbac.authorization.k8s.io
    resources:
      - services # suggestion-api and earlystopping-api
      - events # katib cries is you remove it for unknown reasons
      - persistentvolumeclaims # maybe katib creates volumes, i am not sure
      - deployments # the main deployment that serves suggestion-api and earlystopping-api
      - jobs # trials are executed via jobs
      - cronjobs # i do not know why it is there. maybe you can delay trials
      - serviceaccounts # earlystopping
      - roles # earlystopping
      - rolebindings # earlystopping
    verbs:
      - get
      - list
      - watch
      - patch
      - create
  - apiGroups:
      - ""
    resources:
      - configmaps # i do not have a clue
      - persistentvolumes # i do not have a clue
      - namespaces 
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - kubeflow.org
    resources: # everything here is obvious
      - experiments
      - experiments/status
      - experiments/finalizers
      - trials
      - trials/status
      - trials/finalizers
      - suggestions
      - suggestions/status
      - suggestions/finalizers
      - tfjobs
      - pytorchjobs
      - mpijobs
      - xgboostjobs
      - mxjobs
    verbs:
      - "*"

**Comment by maanur** [comment]
We use Katib with customized RBAC in a desperate attempt to fit in corporate information security requirements.
Our ClusterRoles are separated: the first declares cluster-level common resources access, the second declares kubeflow.org-only cluster-level access, and the third declares namespace-level rules.

I've never posted it here, because I believe that RBAC manifests should be autogenerated from in-code markings, for example, by kubebuilder.
[rbac.yaml.zip](https://github.com/kubeflow/katib/files/8161735/rbac.yaml.zip)

Take a look, please. @juliusvonkohout  @andreyvelich

**Comment by johnugeorge** [comment]
@maanur Can you create a PR which would be easy to review?

**Comment by maanur** [comment]
> Can you create a PR which would be easy to review?
Ok, I'll create PR after adding some comments to the yaml, but I really see it as a temporary solution.

**Comment by johnugeorge** [comment]
I meant,  Do you have a narrowed set of RBAC rules?

**Comment by maanur** [comment]
> Do you have a narrowed set of RBAC rules?

Yes, I have, but I'd better suggest a more sustainable solution.

**Comment by juliusvonkohout** [comment]
> > Do you have a narrowed set of RBAC rules?
> 
> Yes, I have, but I'd better suggest a more sustainable solution.

What do you have in mind?
